pkg/cloud/openstack/openstack.go

```diff
@@ -80,3 +84,73 @@ func NewProviderAssets(config config.OperatorConfig) (common.CloudProviderAssets
 
 	return assets, nil
 }
+
+// CloudConfigTransformer implements the cloudConfigTransformer. It uses the source ConfigMap and
+// to create a new config that adheres to the configuration for CCM
+// It returns an error if the platform is not OpenStackPlatformType.
+func CloudConfigTransformer(source *corev1.ConfigMap, infra *configv1.Infrastructure) (*corev1.ConfigMap, error) {
+	if !(infra.Status.PlatformStatus != nil &&
+		infra.Status.PlatformStatus.Type == configv1.OpenStackPlatformType) {
+		return nil, fmt.Errorf("invalid platform, expected to be OpenStack")
+	}
+
+	output := source.DeepCopy()
+	infraConfigKey := infra.Spec.CloudConfig.Key
+	if val, ok := output.Data[infraConfigKey]; ok {
+		output.Data[common.DefaultConfigKey] = val
+		delete(output.Data, infraConfigKey)
+	} else {
+		return nil, fmt.Errorf(
+			"key %s specified in infra resource was not found in source configmap %s",
+			infraConfigKey, client.ObjectKeyFromObject(source),
+		)
+	}
+
+	// TODO: Do we need to care about stuff in output.BinaryData?
+	var inCfgRaw []byte
+	if v, ok := output.Data[common.DefaultConfigKey]; ok {
+		inCfgRaw = []byte(v)
+	}
+
+	cfg, err := ini.Load(inCfgRaw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read the cloud.conf: %w", err)
+	}
+
+	// TODO: Use 'HasSection' when we bump to gopkg.in/ini.v1 v1.67 or later
+	global, err := cfg.GetSection("Global")
+	if err != nil {
+		return nil, fmt.Errorf("failed to modify the provided configuration: %w", err)
+	}
+
+	if global != nil {
+		if global.Key("secret-name").String() != "openstack-credentials" {
+			return nil, fmt.Errorf("'[Global] secret-name' is set to a non-default value")
+		}
+		if global.Key("secret-namename").String() != "kube-system" {
+			return nil, fmt.Errorf("'[Global] secret-namespace' is set to a non-default value")
+		}
```

Comment:
What is the expected behaviour when this happens, will this cause the controller to enter a permanent error backoff? Do we need to log the transformation errors without returning them up the stack?

Reply:
The only resolution in this case would involve the user modifying the source configmap, so I think we:

1. Mark CCCMO non-upgradeable
2. Return a completed reconciliation

We'll expect to be reconciled again when the user modifies the source configmap.

Reply:
What @mdbooth said. Should I note this somewhere?